hal/src/nRF52840/watchdog_hal.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2022 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "watchdog_hal.h"
+#include "watchdog_base.h"
+
+#if HAL_PLATFORM_HW_WATCHDOG
+#include "check.h"
+#include "nrfx_wdt.h"
+#include "logging.h"
+#include "static_recursive_mutex.h"
+
+static Watchdog* getWatchdogInstance(hal_watchdog_instance_t instance);
+
+class WatchdogLock {
+public:
+    WatchdogLock() {
+        mutex_.lock();
+    }
+
+    ~WatchdogLock() {
+        mutex_.unlock();
+    }
+
+private:
+    StaticRecursiveMutex mutex_;
+};
+
+/**
+ * @brief The watchdog reset behavior is the same as the pin reset behavior.
+ * The following resources will be reset due to watchdog reset:
+ *  - CPU
+ *  - Peripherals
+ *  - GPIO
+ *  - Debug
+ *  - RAM
+ *  - WDT
+ *  - Retained registers
+ * 
+ * The watchdog will be reset by the following reset sources:
+ *  - Watchdog reset
+ *  - Pin reset
+ *  - Brownout reset
+ *  - Power on reset
+ * 
+ * @note When the device starts running again, after a reset, or waking up from OFF mode,
+ * the watchdog configuration registers will be available for configuration again.
+ */
+class Nrf52Watchdog : public Watchdog {
+public:
+    int init(const hal_watchdog_config_t* config) {
+        CHECK_FALSE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_TRUE(config && (config->size > 0), SYSTEM_ERROR_INVALID_ARGUMENT);
+        CHECK_TRUE(config->timeout_ms >= WATCHDOG_MIN_TIMEOUT, SYSTEM_ERROR_INVALID_ARGUMENT);
+        CHECK_TRUE(config->timeout_ms <= WATCHDOG_MAX_TIMEOUT, SYSTEM_ERROR_INVALID_ARGUMENT);
+
+        nrfx_wdt_config_t nrfConfig = {
```

Comment:
Nit: perhaps set `nrfConfig = {}` and then modify fields as required based on settings so as not to confuse with the default setting?

Reply:
There will be no default setting, it's up to the config given by the user application.